lighthouse-core/gather/gatherers/source-maps.js

```diff
@@ -0,0 +1,149 @@
+ */
+'use strict';
+
+const Gatherer = require('./gatherer.js');
+const Driver = require('../driver.js'); // eslint-disable-line no-unused-vars
+const LHError = require('../../lib/lh-error.js');
+
+/* global fetch */
+
+/**
+ * This function is careful not to parse the response as JSON, as it will
+ * just need to be serialized again over the protocol, and source maps can
+ * be huge.
+ *
+ * If an error occurs, the first character is '!'.
+ *
+ * @param {string[]} urls
+ * @return {Promise<string[]>}
+ */
+/* istanbul ignore next */
+async function fetchSourceMaps(urls) {
+  const responses = urls.map(async (url) => {
+    try {
+      const response = await fetch(url);
+      return response.text();
+    } catch (err) {
+      return '!' + err.toString();
+    }
+  });
+  return Promise.all(responses);
+}
+
+/**
+ * @fileoverview Gets JavaScript source maps.
+ */
+class SourceMaps extends Gatherer {
+  constructor() {
+    super();
+    /** @type {LH.Crdp.Debugger.ScriptParsedEvent[]} */
+    this._scriptParsedEvents = [];
+  }
+
+  /**
+   * @param {Driver} driver
+   * @param {Array<{url: string, sourceMapURL: string}>} sourceMapsToFetch
+   * @return {Promise<LH.Artifacts.SourceMap[]>}
+   */
+  async fetchSourceMapsInPage(driver, sourceMapsToFetch) {
+    const sourceMapUrlsParam = JSON.stringify(sourceMapsToFetch.map(obj => obj.sourceMapURL));
+    // TODO: change default protocol timeout?
+    // driver.setNextProtocolTimeout(250);
+    /** @type {string[]} */
+    const sourceMapJsons =
+      await driver.evaluateAsync(`(${fetchSourceMaps.toString()})(${sourceMapUrlsParam})`);
+
+    /** @type {LH.Artifacts.SourceMap[]} */
+    const sourceMaps = [];
+    for (const [i, json] of sourceMapJsons.entries()) {
+      const url = sourceMapsToFetch[i].url;
+      if (json.startsWith('!')) {
+        sourceMaps.push({url, error: json.substring(1)});
+      } else {
+        sourceMaps.push(this.parseSourceMapJson(url, json));
+      }
+    }
+    return sourceMaps;
+  }
+
+  /**
+   * @param {string} url
+   * @param {string} json
+   * @return {LH.Artifacts.SourceMap}
+   */
+  parseSourceMapJson(url, json) {
+    try {
+      return {
+        url,
+        map: JSON.parse(json),
+      };
+    } catch (err) {
+      // Without this catch, this silently fails and the gatherer returns an empty object... why no visible error?
+      return {url, error: err.toString()};
+    }
+  }
+
+  /**
+   * @param {LH.Crdp.Debugger.ScriptParsedEvent} event
+   */
+  onScriptParsed(event) {
+    this._scriptParsedEvents.push(event);
+  }
+
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   */
+  async beforePass(passContext) {
+    const driver = passContext.driver;
+    driver.on('Debugger.scriptParsed', this.onScriptParsed.bind(this));
```

Comment:
do we need to bind this in the constructor for `.off` to work?

Reply:
Probably ..